tests/ConnectDisconnectTest.test.nut

```diff
@@ -0,0 +1,211 @@
+*
+*/
+  function testDoubleConnectAsync() {
+    assertTrue(server.isconnected(), "Failed pre-run saanity check");
+    return Promise(function(resolve, reject) {
+      cm.connect();
+      assertTrue(server.isconnected(), "should be connected!");
+      assertTrue(cm.isConnected(), "CM should report connected state!");
+
+      cm.onConnect(function() {
+        assertTrue(server.isconnected(), "should be connected!");
+        assertTrue(cm.isConnected(), "CM should report connected state!");
+        resolve();
+      }.bindenv(this));
+
+      cm.connect();
+      assertTrue(server.isconnected(), "should be connected!");
+      assertTrue(cm.isConnected(), "CM should report connected state!");
+    }.bindenv(this))
+    .then(function(val) {
+        assertTrue(cm.isConnected(), "CM should report state as connected!");
+        assertTrue(server.isconnected(), "should be connected again!");
+        _resetCM.call(this);
+    }.bindenv(this))
+    .fail(function(valueOrReason) {
+        _resetCM.call(this);
+        throw valueOrReason;
+      }.bindenv(this));
+  }
+
+  function _checkConnectForCommon(needDisconnect = false) {
+    local flag = 0;
+    assertTrue(server.isconnected(), "Failed pre-run saanity check");
+    if (needDisconnect) {
+      cm.disconnect();
+      assertTrue(!server.isconnected(), "should NOT be connected!");
+    }
+    return Promise(function(resolve, reject) {
+      cm.onDisconnect(function(expected) {
+        assertTrue(!cm.isConnected(), "should NOT be connected!");
+        assertTrue(!server.isconnected(), "should NOT be connected!");
+
+        cm.onConnect(function() {
+          assertTrue(flag==2, "flag should be 2 now");
+          flag++;
+          resolve();
+        }.bindenv(this));
+        flag++;
+        cm.connect();
+      }.bindenv(this));
+
+      cm.connectFor(function() {
+        assertTrue(server.isconnected(), "inside connectFor: should be connected!");
+        flag++;
+      }.bindenv(this));
+    }.bindenv(this))
+    .then(function(val) {
+        assertTrue(cm.isConnected(), "CM should report state as connected!");
+        assertTrue(server.isconnected(), "should be connected again!");
+        assertTrue(flag==3, "flag should be 3 now");
+        _resetCM.call(this);
+    }.bindenv(this))
+    .fail(function(valueOrReason) {
+        _resetCM.call(this);
+        throw valueOrReason;
+      }.bindenv(this));
+  }
+
+  /*
+  * sets connectFor callback calls cm.disconnect before
+  *
+  */
+  function testConnectForFromDisconnectedStateAsync() {
+    return _checkConnectForCommon(true);
+  }
+
+  /*
+  * sets connectFor callback without calling cm.disconnect before
+  *
+  */
+  function testConnectForAsync() {
+    return _checkConnectForCommon(false);
+  }
+
+  function tearDown() {
+    _resetCM.call(this);
+    return "Test finished";
+  }
+
+  /*
+  * resets Cm state to default ones
+  *
+  */
+  function _resetCM() {
```

Comment:
Code duplication. Could you please make a base test class and move the "setUp" and "_reset" methods into the base class ?